macOS: Change drag copy shortcut from Command to Option

tresf opened the pull request:
Closes #2890

michaelgregorius commented on src/gui/AutomatableModelView.cpp:
```diff
@@ -171,7 +171,12 @@ void AutomatableModelView::unsetModel()
+		event->modifiers() & Qt::AltModifier )
+#else
+		event->modifiers() & Qt::ControlModifier )
+#endif
```
This code is repeated several times throughout the code. I propose to put it into a helper method with a speaking name. Doing so would also reduce the places where the `#ifdef LMMS_BUILD_APPLE` checks are needed/done.

tresf replied:
I invite anyone with better understanding of the code to do so and push directly to this branch. ♥️

michaelgregorius replied:
@tresf, I have merged with master to fix a merge conflict and will try to implement what I had in mind in your branch.

michaelgregorius replied:
> I invite anyone with better understanding of the code to do so and push directly to this branch. ♥️

@tresf, I have pushed my changes with several new commits. They mainly introduce the new file `OperatingSystemHelper.h` which contains OS-specific code. Most of the ifdefs of this PR are now "confined" to this file.

I am still a bit skeptical because I would think that most of this should be handled by Qt.

In a "quid pro quo" fashion I would greatly appreciate if someone could review my PR for the faders with a dB scale (which I don't want to link here because the issues are not related). :wink:

tresf replied:
> @tresf, I have pushed my changes with several new commits. They mainly introduce the new file OperatingSystemHelper.h which contains OS-specific code. Most of the ifdefs of this PR are now "confined" to this file.

Thanks.

> I am still a bit skeptical because I would think that most of this should be handled by Qt.

I think this is partially rehashing a conversation here: https://github.com/LMMS/lmms/issues/2890#issuecomment-2174229974.  The documentation specifically discusses the use-case we're using where `mousePressEvent()` and `mouseMoveEvent()` get in the way. 

This is conflated by the fact that "Control" is a mouse modifier for "Copy" on Windows but a keyboard shortcut (e.g. Ctrl + Z) too.

On Mac, `QDrag` sadly doesn't allow "Command" to be a mouse modifier for "Copy", despite it being a keyboard shortcut (e.g. Command + Z) so we're forced to split up the code to differentiate the two, possible a justification for refactoring these variables.

So, I have to somewhat disagree with skepticism to the contrary, because we use the custom mouse events to justify the additional guards, and historically have made "Control" ambiguous, which can't be assumed on Mac.

Rossmaxx commented on include/lmms_basics.h:
```diff
@@ -141,6 +141,12 @@ constexpr const char* UI_CTRL_KEY =
 #else
 "Ctrl";
 #endif
+constexpr const char* UI_ALT_KEY =
```
Just so you note, I'm planning to remove this entire file eventually, so would be good if we find a good place in case i go through it.

tresf replied:
Thanks for the heads up, however I would expect this to be handled as part of that removal (i.e. "refactor").

michaelgregorius replied:
I know this thread is resolved, but just as a heads up: This code is now out of `lmms_basics.h` and in `OperatingSystemHelper.h`.

Rossmaxx replied:
Update: I've took the first step towards removal/scoping down in #7677. After this PR goes through, I'll restrict the usage of `lmms_basics.h` to type declaration.

tresf commented on src/gui/clips/ClipView.cpp:
```diff
@@ -726,7 +727,7 @@ void ClipView::mousePressEvent( QMouseEvent * me )
 						? tr( "Press <%1> and drag to make a copy." )
 						: tr( "Press <%1> for free resizing." );
-			m_hint = TextFloat::displayMessage( tr( "Hint" ), hint.arg(UI_CTRL_KEY),
+			m_hint = TextFloat::displayMessage( tr( "Hint" ), getOSSppecificModifierKeyString(),
```
I think this accidentally blew away the tooltip. 
```suggestion
			m_hint = TextFloat::displayMessage( tr( "Hint" ), hint.arg(getOSSppecificModifierKeyString()),
```

Rossmaxx replied:
You can commit too right?

tresf commented on include/OperatingSystemHelpers.h:
```diff
@@ -0,0 +1,78 @@
+#endif
+}
+
+constexpr Qt::KeyboardModifier getOSSpecificModifierKey()
```
Curious, why is this and `getOSSppecificModifierKeyString` are functions?

Rossmaxx replied:
Might be to simplify the code and reduce the usage of macros to this one file, which is a good way of reducing macros, and some verbosity here.

a better name for this function shall be `getCopyModifier()` similar to the naming in my previous comment.

michaelgregorius replied:
> Curious, why are this and `getOSSppecificModifierKeyString` functions?

I think it was mainly because all the `ifdef`s were previously scattered in the "active" code and therefore my first instinct was to move it to a single place but still with a "functiony" mindset. Using just `constexpr` variables as done by your further changes is the better and less complex solution though.

tresf commented on include/OperatingSystemHelpers.h:
```diff
@@ -0,0 +1,78 @@
+"Alt";
+#endif
+
+constexpr const char* getOSSppecificModifierKeyString()
```
Any ideas for a better name here?  It's predominantly used for copying patterns and linking automation controls.  I'm fine with something like `copyModifierKeyString`, but I think I will quickly forget the currently proposed name(s).

The edge-case here I think is the "free resizing" -- or in other apps "disable snap-to-grid' which is ironically "Control" in Garageband.  I'm curious what the Windows equivalent is.  From what I've read the "Alt" key is the shortcut that PowerPoint and window snapping uses to disable this behavior.  This suggests that it should be split off into its own in the future, reinforcing the name of `copyModifier` for coping and linking, as "freeResizeModifier" would replace the secondary function.

Rossmaxx replied:
`getCopyModifierString()` sounds like a good name for this function, like tresf says.

michaelgregorius replied:
Sorry for the late replies. I was mainly concentrating on some other open PRs and now I am (too) late to the party. Naming definitively is hard but it seems that a good name has been found. :sweat_smile:

michaelgregorius commented:
I am out here now. If someone wants to continue with this (naming, constexpr values instead of functions) feel free to do so.

tresf commented:
> I am out here now. If someone wants to continue with this (naming, constexpr values instead of functions) feel free to do so.

No worries, the code you provided is easy enough to refactor, thanks for the example.  Although I would still like some feedback on my questions, these may be better suited in the form of a commit.  I'll work up something that addresses the comments above.

Rossmaxx commented on src/gui/widgets/Fader.cpp:
```diff
@@ -125,7 +125,8 @@ void Fader::mouseMoveEvent(QMouseEvent* mouseEvent)
 	if (mouseEvent->button() == Qt::LeftButton &&
-			!(mouseEvent->modifiers() & Qt::ControlModifier))
+			!(mouseEvent->modifiers() & Qt::ControlModifier) &&
+			!(mouseEvent->modifiers() & Qt::AltModifier))
```
Would adding an alt modifier here, and in other lines below change behaviour?

Rossmaxx reviewed (commented):
Some thoughts from my side. I could appreciate if we get these PRs through faster.

tresf commented:
Ready for review.

messmerd commented on include/PluginFactory.h:
```diff
@@ -106,6 +106,13 @@ public slots:
 	static std::unique_ptr<PluginFactory> s_instance;
 };
 
+constexpr char LADSPA_PATH_SEPERATOR =
```
This is currently only used in LadspaManager.cpp, but in the CLAP PR I also use it. Maybe it could go in `lmms_constants.h` and be renamed to `PathSeparator`?

Rossmaxx replied:
Isn't PluginFactory a good enough place?

messmerd replied:
No, it's a general-purpose constant, so `lmms_constants.h` is the perfect place for it

tresf replied:
> This is currently only used in LadspaManager.cpp, but in the CLAP PR I also use it. Maybe it could go in `lmms_constants.h` and be renamed to `PathSeparator`?

This is exactly Qt's [`QDir::listSeparator()`](https://doc.qt.io/qt-6/qdir.html#listSeparator), so one option is to simply remove it.

This change was never intended for this code and was changed when ross asked mg to do so, which I believed at the time didn't belong in this PR and I still believe that.  I'd happily remove it.  I kept it out of politeness to ross, but it has absolutely nothing to do this with PR.

My justification for moving it to PluginFactory:
- It's strictly linked to plugin loading
- It'll likely be needed in the future when we add support for multiple path locations for other plugins types (e.g. VST, LV2, etc)

messmerd replied:
I didn't know Qt had `QDir::listSeparator()` for that. Still, I'd rather we use our own because it doesn't require including a Qt header.

I don't think it's right for PluginFactory because it's generally useful, and if I'm a dev and need a common constant like that, I would think "let's check lmms_constants.h" not "let's check PluginFactory.h".

tresf replied:
I've happily removed it from this PR.

When the team decides to refactor it, if the project decides to keep it, it'll need a new name anyway.  PluginFactory.h is already included in plugins so it was a free dependency and also one that was explicitly included, unlike its current usage.

tresf replied:
> I'd rather we use our own because it doesn't require including a Qt header.

I had a feeling this would be the preferred route.  I hope you don't mind me pushing this one off though, especially considering there's at least one open PR that depends on it.

messmerd commented on include/KeyboardShortcuts.h:
```diff
@@ -0,0 +1,77 @@
+ * KeyboardShortcuts.h - Cross-platform handling of keyboard modifier keys
+ *
+ * Copyright (c) 2025 Michael Gregorius
+ * Copyright (c) 2015 Tres Finocchiaro <tres.finocchiaro/at/gmail.com>
```
```suggestion
 * Copyright (c) 2025 Tres Finocchiaro <tres.finocchiaro/at/gmail.com>
```
Hello, I'm from the future

messmerd reviewed (approved):
I don't have macOS to test it, but it looks fine

tresf commented:
> I don't have macOS to test it, but it looks fine

I tested copying, linking and also ensure the negations worked (e.g. could still adjust things).  It's so very nice to have copy shortcuts back.
